src/OpenApi/src/Transformers/TypeBasedOpenApiDocumentTransformer.cs

```diff
@@ -0,0 +1,34 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.AspNetCore.OpenApi;
+
+internal sealed class TypeBasedOpenApiDocumentTransformer(Type transformerType) : IOpenApiDocumentTransformer
+{
+    private readonly ObjectFactory _transformerFactory = ActivatorUtilities.CreateFactory(transformerType, []);
+    private IOpenApiDocumentTransformer? _transformer;
+
+    public ValueTask DisposeAsync()
+    {
+        if (_transformer is IAsyncDisposable asyncDisposable)
+        {
+            return asyncDisposable.DisposeAsync();
+        }
+        if (_transformer is IDisposable disposable)
+        {
+            disposable.Dispose();
+        }
+        return ValueTask.CompletedTask;
+    }
+
+    public Task TransformAsync(OpenApiDocument document, OpenApiDocumentTransformerContext context, CancellationToken cancellationToken)
+    {
+        _transformer = _transformerFactory.Invoke(context.ApplicationServices, []) as IOpenApiDocumentTransformer;
```

Comment:
This class can be hit in parallel by parallel requests, so the instance in the `_transformer` field isn't going to necessarily be preserved by the time `DisposeAsync` is called. You could just do the dispose inline and get rid of `_transformer`.